### PR TITLE
Create phi3mini.py

### PR DESCRIPTION
**Pull request by mayani-nv**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by cuichenx — COMMENTED** [review]
This file looks good to me. Can you expose the config and model classes in this file to various __init__.py? See other models for an example. The goal is to be able to do `llm.Phi3Model`

File: nemo/collections/llm/recipes/phi3_mini_4k_instruct.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+    if performance_mode:
+        recipe = pretrain_performance_optimizations(recipe)
+
+    return recipe
+
+
+def pretrain_performance_optimizations(recipe: run.Partial) -> run.Partial:
+    """
+    Create a performance-optimized pre-training recipe for Llama3 8B model.
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/phi3_mini_4k_instruct.py]: same for this

File: nemo/collections/llm/recipes/phi3_mini_4k_instruct.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            callbacks=[run.Config(TimingCallback)],
+        ),
+        data=run.Config(MockDataModule, seq_length=4096, global_batch_size=512, micro_batch_size=1),
+        log=default_log(dir=dir, name=name, tensorboard_logger=tensorboard_logger(name=name)),
+        optim=distributed_fused_adam_with_cosine_annealing(max_lr=3e-4),
+        resume=default_resume(),
+    )
+
+    if performance_mode:
+        recipe = pretrain_performance_optimizations(recipe)
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/phi3_mini_4k_instruct.py:201]: this should be removed

File: nemo/collections/llm/recipes/phi3_mini_4k_instruct.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    # Sequence length settings in the model and dataset must agree
+    recipe.model.config.seq_length = seq_length
+    recipe.data.seq_length = seq_length
+    if packed_sequence:
+        recipe.data.pad_to_max_length = True
+        recipe.data.packed_sequence_specs = run.Config(PackedSequenceSpecs, packed_sequence_size=seq_length)
+
+    if performance_mode:
+        recipe = finetune_performance_optimizations(recipe, peft_scheme)
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/phi3_mini_4k_instruct.py:289]: this should be removed

Comment (mayani-nv) [reply nemo/collections/llm/recipes/phi3_mini_4k_instruct.py]: done, thanks

File: nemo/collections/llm/recipes/phi3_mini_4k_instruct.py (hunk 2 above)
Comment (mayani-nv) [reply nemo/collections/llm/recipes/phi3_mini_4k_instruct.py]: done, thanks

File: nemo/collections/llm/recipes/phi3_mini_4k_instruct.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    name: str = "default",
+    num_nodes: int = 1,
+    tensor_parallelism: int = 1,
+    num_gpus_per_node: int = 1,
+    max_steps: int = 1168251,
+    performance_mode: bool = False,
+    fn: Callable = pretrain,
+) -> run.Partial:
+    """
+    Create a pre-training recipe for Llama3 8B model.
```
[the remark this answers is not in the document]
Comment (mayani-nv) [reply nemo/collections/llm/recipes/phi3_mini_4k_instruct.py]: done, thanks

File: nemo/collections/llm/__init__.py [hunk 5]
```diff
@@ -143,6 +146,9 @@
     "Nemotron4Config15B",
     "Nemotron4Config340B",
     "NemotronConfig",
+    "Phi3config",
```
[the remark this answers is not in the document]
Comment (mayani-nv) [reply nemo/collections/llm/__init__.py]: removed

**Review by cuichenx — APPROVED** [review]
LGTM

**Pull request merged**